File: src/main/java/com/hedvig/backoffice/services/claims/dto/ClaimPayment.java
package com.hedvig.backoffice.services.claims.dto;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.hedvig.backoffice.services.payments.dto.SelectedPayoutDetails;
import java.time.LocalDateTime;
import java.util.UUID;
import javax.annotation.Nullable;
import javax.money.MonetaryAmount;
import javax.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Data;

@Data
@AllArgsConstructor
public class ClaimPayment {
    @NotNull
    public String claimId;

    @NotNull
    public MonetaryAmount amount;

    @NotNull
    public MonetaryAmount deductible;

    @NotNull
    public String note;

    public boolean exGratia;

    @NotNull
    ClaimPaymentType type;

    @NotNull
    String handlerReference;

    boolean sanctionListSkipped;

    @Nullable
    @JsonProperty("id")
    public String paymentId;

    @JsonProperty("date")
    LocalDateTime registrationDate;

    ClaimPaymentStatus payoutStatus;

    UUID transactionId;

    String carrier;

    SelectedPayoutDetails payoutDetails;

    public ClaimPayment(
        @NotNull String claimId,
        @NotNull MonetaryAmount amount,
        @NotNull MonetaryAmount deductible,
        @NotNull String note,
        boolean exGratia,
        @NotNull ClaimPaymentType type,
        @NotNull String handlerReference,
        boolean sanctionListSkipped,
        @NotNull String carrier,
        SelectedPayoutDetails payoutDetails
    ) {
        this.claimId = claimId;
        this.amount = amount;
        this.deductible = deductible;
        this.note = note;
        this.exGratia = exGratia;
        this.type = type;
        this.handlerReference = handlerReference;
        this.sanctionListSkipped = sanctionListSkipped;
        this.carrier = carrier;
        this.payoutDetails = payoutDetails;
    }
}

File: src/main/java/com/hedvig/backoffice/services/claims/dto/ClaimPayment.kt
package com.hedvig.backoffice.services.claims.dto

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.hedvig.backoffice.services.payments.dto.SelectedPayoutDetails
import com.hedvig.backoffice.util.DoubleOrMonetaryAmountToMonetaryAmountDeserializer
import lombok.AllArgsConstructor
import lombok.Data
import java.time.LocalDateTime
import java.util.UUID
import javax.money.MonetaryAmount
import javax.validation.constraints.NotNull

data class ClaimPayment(
    var claimId: String,
    @JsonTypeInfo(use = JsonTypeInfo.Id.NONE) @field:JsonDeserialize(using = DoubleOrMonetaryAmountToMonetaryAmountDeserializer::class) var amount: @NotNull MonetaryAmount?,
    @JsonTypeInfo(use = JsonTypeInfo.Id.NONE) @field:JsonDeserialize(using = DoubleOrMonetaryAmountToMonetaryAmountDeserializer::class) var deductible: @NotNull MonetaryAmount?,
    val note: String,
    val exGratia: Boolean,
    val type: ClaimPaymentType,
    var handlerReference: String,
    val sanctionListSkipped: Boolean,
    val payoutDetails: SelectedPayoutDetails,
    //The following fields are used when getting a claim, and are null when creating a payment:
    @JsonProperty("id")
    var paymentId: String? = null,
    @JsonProperty("date")
    var registrationDate: LocalDateTime? = null,
    var payoutStatus: ClaimPaymentStatus? = null,
    var transactionId: UUID? = null
)
